include/tvm/runtime/memory/memory_manager.h

```diff
@@ -157,7 +160,7 @@ class StorageObj : public Object {
 /*! \brief reference to storage. */
 class Storage : public ObjectRef {
  public:
-  TVM_DLL explicit Storage(Buffer buffer);
+  TVM_DLL explicit Storage(Buffer buffer, Allocator* allocator);
```

Comment:
Is this constructor actually used? In this PR I only see manual attachment like `storage_obj->allocator = allocator;`.

We started to see a segfault like `free(): invalid pointer` when a python process that runs a script exits, after we integrated recent upstream changes. I'm investigating if the segfault is really due to some recent upstream changes, and which one if so.

Reply:
Yes, this is used so recycle goes back to original allocator, in this case we need to recycle to IPCMmeory in the special allocator

Reply:
> Yes, this is used

Can you point out the code? In our fork, I don't see it used outside of `paged_kv_cache.cc`.

Reply:
There are mixture usages of `Storage`/`StorageObj` in TVM. The only use of the constructor seems to be in `paged_kv_cache.cc` in TVM. For other constructions of `Storage`, they all use `make_object` and directly set the allocator
https://github.com/apache/tvm/blob/622bd150dd331780eb41a1c67c65aae802eb9b20/src/runtime/disco/cuda_ipc/cuda_ipc_memory.cc#L204-L210
https://github.com/apache/tvm/blob/622bd150dd331780eb41a1c67c65aae802eb9b20/src/runtime/relax_vm/builtin.cc
https://github.com/apache/tvm/blob/622bd150dd331780eb41a1c67c65aae802eb9b20/src/runtime/vm/vm.cc#L848-L859

There is a use of the constructor on MLC LLM side https://github.com/mlc-ai/mlc-llm/blob/main/cpp/serve/model.cc#L68-L69, where we use the constructor so that we don't need to expand it into the `make_object` lines.

Reply:
What worries me is that the code is doing `allocator->Free(buffer);` in the destructor. But how can you be sure that it's safe if there is no guarantee that the allocator is always attached to the storage when it is created?

Reply:
I see. Can we just check the allocator is defined by adding ICHECK in the destructor, or change all `make_object` to use the constructor and add ICHECK inside? If I remember correctly, I made a pass over all occurrences and attached the allocators.

Reply:
> Can we just check the allocator is defined by adding ICHECK in the destructor

Yes, I tried that in our fork but the segfault is still there. So this PR is not a cause of our problem.

But still, this doesn't change my concern about this code. I'm sure that you checked all usages, but there is always a possibility to miss some. And you never know about the usages in new code or someone else's fork.

Reply:
Yeah I agree with your concern. I can send a patch to add some checks for more safety later on.